src/main/java/net/minecraftforge/client/gui/tabs/GuiTab.java

```diff
@@ -0,0 +1,318 @@
+    {
+        return (getIndex() > 11 ? (getIndex() - 12) % 12 : this.getIndex()) % 6;
+    }
+
+    /**
+     * @return Whether this tab's parent GUI is the current GUI
+     */
+    public boolean isActiveTab(Class<? extends GuiContainer> guiContainer)
+    {
+        return guiContainer == getTargetGui();
+    }
+
+    /**
+     * @return The tab index / 12
+     */
+    public int getPage()
+    {
+        return getIndex() / 12;
+    }
+
+    /**
+     * @return The unlocalized name for this tab
+     */
+    public String getUnlocalizedName()
+    {
+        return "guiGroup." + name;
+    }
+
+    public ResourceLocation getIconResLoc()
+    {
+        return iconResLoc;
+    }
+
+    public ItemStack getIconStack()
+    {
+        return iconStack;
+    }
+
+    /**
+     * @return A list of the appended tabs for the given GuiContainer
+     */
+    public static List<GuiTab> getTabsForGui(Class<? extends GuiContainer> guiContainer)
+    {
+        return tabRegistry.get(guiContainer);
+    }
+
+    /**
+     * Adds the GuiTab to GuiContainer
+     *
+     * @param parentContainer The GuiContainer to add a tab too
+     * @return The GuiTab
+     */
+    public GuiTab addTo(Class<? extends GuiContainer> parentContainer)
+    {
+        if (parentContainer == GuiContainerCreative.class)
+        {
+            throw new IllegalArgumentException("Cannot add tabs to creative inventory. Use CreativeTabs instead");
+        }
+        else if (parentContainer == InventoryEffectRenderer.class)
+        {
+            throw new IllegalArgumentException(name + " tab has a parent that is ambiguous with the creative inventory. cannot proceed with registry.");
+        }
+
+        this.parentContainer = parentContainer;
+        tabRegistry.put(parentContainer, this);
+        return this;
+    }
+
+    /**
+     * Adds the GuiTab to GuiContainers
+     *
+     * @param parentContainers The GuiContainers to add a tab too
+     * @return The GuiTab
+     */
+    public GuiTab addTo(Class<? extends GuiContainer>... parentContainers)
+    {
+        Arrays.stream(parentContainers).forEach(this::addTo);
+        return this;
+    }
+
+    /**
+     * Sets the GuiContainer this tab should show on as the main tab.
+     *
+     * @param targetGui The target gui to shown on
+     * @return The GuiTab instance
+     */
+    public GuiTab setTargetGui(Class<? extends GuiContainer> targetGui)
+    {
+        this.targetGui = targetGui;
+        return this;
+    }
+
+    /**
+     * Process interaction when the tab is clicked. should be used to send a packet to the server and open a GUI
+     *
+     * @param guiContainer The parent gui
+     */
+    public void onTabClicked(GuiContainer guiContainer)
+    {
+        List<GuiTab> targetTabs = getTabsForGui(targetGui);
+        List<GuiTab> tabs = getTabsForGui(guiContainer.getClass());
+        tabs.stream().filter(Predicates.not(targetTabs::contains)).forEach(t -> t.addTo(targetGui));
```

Comment:
What's happening here? From the JavaDoc it sounds like the method should be abstract, and I just cannot figure out what the implementation should be doing.

Reply:
I'm not 100% sure, but when I had to use this code I had to call super on the test mod calls, I think it's because it calls the different types of tabs in the target GUI and syncs them with client /server events.

Reply:
@HenryLoenwind example of what the implementation should be doing: https://github.com/sokratis12GR/MinecraftForge/blob/42acdbc4d82d8e6cfde13aa3ec2eb504749de30d/src/main/java/net/minecraftforge/client/gui/tabs/DefaultVanillaGuiTabs.java#L71#L75
```JAVA
        @Override
        public void onTabClicked(GuiContainer guiContainer)
        {
            mc.displayGuiScreen(new GuiInventory(mc.player));
        }
```
https://github.com/sokratis12GR/MinecraftForge/blob/42acdbc4d82d8e6cfde13aa3ec2eb504749de30d/src/test/java/net/minecraftforge/test/GuiTabsTest.java#L83
```JAVA
            new GuiTab("Tab1", new ItemStack(Blocks.BEDROCK), GuiChest.class)
            {
                @Override
                public void onTabClicked(GuiContainer guiContainer)
                {
                    super.onTabClicked(guiContainer);
                    GuiTabsTest.INSTANCE.sendToServer(new TabMessageHandler.TabMessage(0));
                }
            }.addTo(TestGuiContainer.class);
```
But after taking a look at it again, I guess it could really be abstract ? I don't really know what would be the best solution for this...

Reply:
My concern is that there's something in the default implementation that I don't fully understand. And then I see tabs that do call super, and tabs that don't do super. Without any further explanation (in the javadoc) people will code by "strategy random". Some will add the super call by habit, others won't. For some their tabs will do what they should, for others they won't. Then there will be countless "howto" guides on how to work around that, all of them wrong.

So, either:

1. all tabs need that code. Make it final and have it call `abstract doOpenTab(...)`` or something.
2. no tab needs it. Then drop it and make the method abstract.
3. some tabs need it. Make a helper method with that code and some javadoc that explains when and why to call it. Make the original method abstract.

Forge's already suffering enough from the "and how do I use that?" plague.

Reply:
Alrighty, will do some further testing, but after some of the initial tests seems like they "none of them need it that are supported by vanilla syncing/packetting".
I guess I'll make it an abstract method and make it an utility/helper method for those which need it.

Reply:
In the first releases, this was the primary argument... 
if I recall correctly : 
Gui's that extend only Gui don't forcibly need a super. you also have no control over where to place the tabs.
Gui's that are containers do need super, and have a predefined space on the screen where the image of the container is drawn. 
The opinions where 50/50 on wether is should be for the very base of gui's, the Gui class, or from GuiContainer and it's children. 

The sole reason I made this PR in the very beginning, was to have more people use a tab in the __player inventory__, just like a lot of mods (used to) do. I was interested (as end user of forge) in using a simple method to add a tab, and __not create compatibility issues__ between mods. 

I do understand the principles of Forge, and therefor the bigger implementation it needs. 

note : today, i'm probably behind on how the code works internally. Don't want to create a fuss.

What I really want to say here, is that all of the three points listed by HenryLoewind are all equally viable. You might have a hard time deciding on what would be the correct one. 
Don't overthink it too much, and go with what you think is best ! 

thanks again sokratis, I believe in you, you can get this PR trough too the forge API !!!

Reply:
@ArtixAllMighty sorry for pinging you, but why as far as I understand this code is being required by GuiContainer in order to have corrected position for the GuiTabs per each GuiContainer ?
If so, I guess it will be required by all GuiContainers to call super on it, so no issues appear ?

Reply:
yes. the super call would be needed in that case. GuiCOntainers have variables to define the background image's location to allow it being centered.

Reply:
Alrighty, so all gui tabs using it would *require* to call on super